### ISSUE TITLE
Java 14 SwitchExpression Check Support validation: FinalLocalVariableCheck

### ISSUE THREAD

**Issue by nmancus1** [checkstyle/checkstyle#8668]
Child of #8658 
Check documentation: https://checkstyle.sourceforge.io/config_coding.html#FinalLocalVariable

_Checks that local variables that never have their values changed are declared final. The check can be configured to also check that unchanged parameters are declared final._

```
➜  src /usr/lib/jvm/java-14-openjdk/bin/javac --enable-preview --source 14 TestClass.java                               
➜  src cat config.xml                                                                                                   
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="FinalLocalVariable"/>
    </module>
</module>
➜  src cat TestClass.java                                                                                               
public class TestClass {

    int howMany1(int k) {
        int x = 2; // ok
        int number = 42; // violation
        switch (k) {
            case 1:
                int y = x + number; // violation
                x = y;
                break;
            case 2:
                int otherNumber = 42; // violation
                int z = x + 2; // violation
                x = z;
                break;
            case 3:
                int a = x + 3; // violation
                x = a;
                break;
            default:
                int b = x + 4; // violation
                x = b;
        }
        return x;
    }

    int howMany2(int k) {
        int x = 2; // violation
        int number = 42; // violation
        return switch (k) {
            case 1 -> {
                int y = x + number; // violation
                yield y;
            }
            case 2 -> {
                int otherInt = 42; // violation
                int z = x + 2; // violation
                final int finalInt = 43;
                yield z;
            }
            case 3 -> {
                int anotherInt = 41; // violation
                int a = x + 3; // violation
                yield a;
            }
            default -> {
                int thisOtherInt = 99; // violation
                int b = x + 4; // violation
                yield b;
            }
        };
    }

    int howMany3(int x) {
        return switch (x) {
            case 1:
                int number = 55; // violation
                int y = x + number; // violation
                final int finalInt = 2; // ok
                yield y;
            case 2:
                int otherNumber = 42; // violation
                int z = x + 2; // violation
                x = z;
                yield otherNumber;
            case 3:
                int a = x + 3; // violation
                x = a;
                yield a;
            default:
                int b = x + 4; // violation
                x = b;
                yield b;
        };
    }

}
➜  src java -jar /home/nick/IdeaProjects/checkstyle/target/checkstyle-8.36-SNAPSHOT-all.jar -c config.xml TestClass.java
Starting audit...
[ERROR] /home/nick/Desktop/enhanced-switch/src/TestClass.java:5:13: Variable 'number' should be declared final. [FinalLocalVariable]
[ERROR] /home/nick/Desktop/enhanced-switch/src/TestClass.java:8:21: Variable 'y' should be declared final. [FinalLocalVariable]
[ERROR] /home/nick/Desktop/enhanced-switch/src/TestClass.java:12:21: Variable 'otherNumber' should be declared final. [FinalLocalVariable]
[ERROR] /home/nick/Desktop/enhanced-switch/src/TestClass.java:13:21: Variable 'z' should be declared final. [FinalLocalVariable]
[ERROR] /home/nick/Desktop/enhanced-switch/src/TestClass.java:17:21: Variable 'a' should be declared final. [FinalLocalVariable]
[ERROR] /home/nick/Desktop/enhanced-switch/src/TestClass.java:21:21: Variable 'b' should be declared final. [FinalLocalVariable]
[ERROR] /home/nick/Desktop/enhanced-switch/src/TestClass.java:28:13: Variable 'x' should be declared final. [FinalLocalVariable]
[ERROR] /home/nick/Desktop/enhanced-switch/src/TestClass.java:29:13: Variable 'number' should be declared final. [FinalLocalVariable]
[ERROR] /home/nick/Desktop/enhanced-switch/src/TestClass.java:32:21: Variable 'y' should be declared final. [FinalLocalVariable]
[ERROR] /home/nick/Desktop/enhanced-switch/src/TestClass.java:36:21: Variable 'otherInt' should be declared final. [FinalLocalVariable]
[ERROR] /home/nick/Desktop/enhanced-switch/src/TestClass.java:37:21: Variable 'z' should be declared final. [FinalLocalVariable]
[ERROR] /home/nick/Desktop/enhanced-switch/src/TestClass.java:42:21: Variable 'anotherInt' should be declared final. [FinalLocalVariable]
[ERROR] /home/nick/Desktop/enhanced-switch/src/TestClass.java:43:21: Variable 'a' should be declared final. [FinalLocalVariable]
[ERROR] /home/nick/Desktop/enhanced-switch/src/TestClass.java:47:21: Variable 'thisOtherInt' should be declared final. [FinalLocalVariable]
[ERROR] /home/nick/Desktop/enhanced-switch/src/TestClass.java:48:21: Variable 'b' should be declared final. [FinalLocalVariable]
[ERROR] /home/nick/Desktop/enhanced-switch/src/TestClass.java:57:21: Variable 'number' should be declared final. [FinalLocalVariable]
[ERROR] /home/nick/Desktop/enhanced-switch/src/TestClass.java:58:21: Variable 'y' should be declared final. [FinalLocalVariable]
[ERROR] /home/nick/Desktop/enhanced-switch/src/TestClass.java:62:21: Variable 'otherNumber' should be declared final. [FinalLocalVariable]
[ERROR] /home/nick/Desktop/enhanced-switch/src/TestClass.java:63:21: Variable 'z' should be declared final. [FinalLocalVariable]
[ERROR] /home/nick/Desktop/enhanced-switch/src/TestClass.java:67:21: Variable 'a' should be declared final. [FinalLocalVariable]
[ERROR] /home/nick/Desktop/enhanced-switch/src/TestClass.java:71:21: Variable 'b' should be declared final. [FinalLocalVariable]
Audit done.
Checkstyle ends with 21 errors.

```

This check still works as intended within switch expressions and switch rules, so I think that we can close this one.

**Comment by pbludov** [comment]
The test case should have an outer variable:
```java
public class TestClass {

    void foo1() throws Exception {

        Exception e; // violation

        final int a = (int) Math.random();
        final int b = (int) Math.random();

        switch (a) {
        case 0:
            e = new Exception();
            break;
        case 1:
            if (b == 0) {
                e = new Exception();
                break;
            }

            if (b == 1) {
                e = new Exception();
            }
            else {
                e = new Exception();
            }
            break;
        case 2:
            if (b == 0) {
                return;
            }

            e = new Exception();
            break;
        default:
            e = new Exception();
            break;
        }

        throw e;
    }

    void foo2() throws Exception {

        Exception e; // should be violation

        final int a = (int) Math.random();
        final int b = (int) Math.random();

        switch (a) {
            case 0 -> {
                e = new Exception();
            }
            case 1 -> {
                if (b == 0) {
                    e = new Exception();
                    break;
                }

                if (b == 1) {
                    e = new Exception();
                }
                else {
                    e = new Exception();
                }
            }
            case 2 -> {
                if (b == 0) {
                    return;
                }

                e = new Exception();
            }
            default -> {
                e = new Exception();
            }
        }

        throw e;
    }

}
```

There should be two violations, but only one reported:
```
Starting audit...
[ERROR] TestClass.java:7:19: Variable 'e' should be declared final. [FinalLocalVariable]
Audit done.
```
